apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ThreadProfiler.java

```diff
@@ -0,0 +1,153 @@
+    // thread dump sequence
+    private int dumpSequence = 0;
+
+    public ThreadProfiler(TracingContext tracingContext, ID traceSegmentId, Thread profilingThread, ProfileTaskExecutionContext executionContext) {
+        this.tracingContext = tracingContext;
+        this.traceSegmentId = traceSegmentId;
+        this.profilingThread = profilingThread;
+        this.executionContext = executionContext;
+        this.profilingMaxTimeMills = TimeUnit.MINUTES.toMillis(Config.Profile.MAX_DURATION);
+    }
+
+    /**
+     * If tracing start time greater than {@link ProfileTask#getMinDurationThreshold()}, then start to profiling trace
+     */
+    public void startProfilingIfNeed() {
+        if (System.currentTimeMillis() - tracingContext.createTime() > executionContext.getTask().getMinDurationThreshold()) {
+            this.profilingStartTime = System.currentTimeMillis();
+            this.profilingStatus = ProfilingStatus.PROFILING;
+        }
+    }
+
+    /**
+     * Stop profiling status
+     */
+    public void stopProfiling() {
+        this.profilingStatus = ProfilingStatus.STOPPED;
+    }
+
+    /**
+     * dump tracing thread and build thread snapshot
+     *
+     * @return snapshot, if null means dump snapshot error, should stop it
+     */
+    public TracingThreadSnapshot buildSnapshot() {
+        if (!isProfilingProfilingContinuable()) {
+            return null;
+        }
+
+        long currentTime = System.currentTimeMillis();
+        // dump thread
+        StackTraceElement[] stackTrace;
+        try {
+            stackTrace = profilingThread.getStackTrace();
+
+            // stack depth is zero, means thread is already run finished
+            if (stackTrace.length == 0) {
+                return null;
+            }
+        } catch (Exception e) {
+            // dump error ignore and make this profiler stop
+            return null;
+        }
+
+        // if is first dump, check is can start profiling
+        if (dumpSequence == 0 && (!executionContext.isStartProfileable())) {
+            return null;
+        }
+
+        int dumpElementCount = Math.min(stackTrace.length, Config.Profile.DUMP_MAX_STACK_DEPTH);
+
+        // use inverted order, because thread dump is start with bottom
+        final ArrayList<String> stackList = new ArrayList<>(dumpElementCount);
+        for (int i = dumpElementCount - 1; i >= 0; i--) {
+            stackList.add(buildStackElementCodeSignature(stackTrace[i]));
+        }
+
+        String taskId = executionContext.getTask().getTaskId();
+        return new TracingThreadSnapshot(taskId, traceSegmentId, dumpSequence++, currentTime, stackList);
+    }
+
+    /**
+     * build thread stack element code signature
+     *
+     * @return code sign: className.methodName:lineNumber
+     */
+    private String buildStackElementCodeSignature(StackTraceElement element) {
+        return element.getClassName() + "." + element.getMethodName() + ":" + element.getLineNumber();
```

Comment:
Maybe take the case when `element.getLineNumber() < 0` into consideration? Make it `Unknown Source`?

Reply:
I am OK with line number = -1. :) Both are fine to me.

Reply:
It's ok to me, "Unknown Source" takes more bandwidth then "-1" after all